zkevm-circuits/src/compress_circuit.rs

```diff
@@ -0,0 +1,498 @@
+    input: Column<Advice>,      // input bitstream byte/row
+    // output parts:
+    // encoded value
+    data: Column<Advice>,
+    data_rlc: Column<Advice>,
+    // encoded bit length
+    num_bits: Column<Advice>,
+    // randomness multiplier selector
+    r_multi: Column<Advice>,
+    // the factor with which the part needs to be multiplied
+    shift_factor: Column<Advice>,
+    // the factor with which the part needs to be multiplied when
+    // we decode it
+    decode_factor: Column<Advice>,
+    // fixed huffman table
+    huffman_table: [TableColumn; 3], // [value, code, code_length]
+
+    _marker: PhantomData<F>,
+}
+
+/// Circuit configuration arguments
+pub struct CompressCircuitConfigArgs<F> {
+    // randomness used to do RLC
+    pub randomness: Expression<F>,
+}
+
+impl<F: Field> SubCircuitConfig<F> for CompressCircuitConfig<F> {
+    type ConfigArgs = CompressCircuitConfigArgs<F>;
+
+    fn new(
+        meta: &mut ConstraintSystem<F>,
+        CompressCircuitConfigArgs { randomness }: Self::ConfigArgs,
+    ) -> Self {
+        let q_enable = meta.selector();
+        let word_begin = meta.advice_column();
+        let r_multi = meta.advice_column();
+        let input = meta.advice_column();
+        let data = meta.advice_column();
+        let data_rlc = meta.advice_column();
+        let num_bits = meta.advice_column();
+        let shift_factor = meta.advice_column();
+        let decode_factor = meta.advice_column();
+        let huffman_table = array_init::array_init(|_| meta.lookup_table_column());
+
+        meta.enable_equality(data_rlc);
+
+        let mut parts = vec![];
+        meta.create_gate("split parts", |meta| {
+            // next 4 rows as 4 parts
+            for i in 0..MAX_PARTS as i32 {
+                let data = meta.query_advice(data, Rotation(i));
+                let num_bits = meta.query_advice(num_bits, Rotation(i));
+                let r_multi = meta.query_advice(r_multi, Rotation(i));
+                let shift_factor = meta.query_advice(shift_factor, Rotation(i));
+                let decode_factor = meta.query_advice(decode_factor, Rotation(i));
+
+                parts.push(Part {
+                    data,
+                    num_bits,
+                    r_multi,
+                    shift_factor,
+                    decode_factor,
+                });
+            }
+            vec![0.expr()]
+        });
+
+        meta.lookup("huffman table", |meta| {
+            // lookups:
+            //  1. input = huffman_value
+            //  2. data0 * decode_factor0 + data1 * decode_factor1 + data2 * decode_factor2
+            //      + data3 * decode_factor3 = huffman_code
+            //  3. num_bits0 + num_bits1 + num_bits2 + num_bits3 = huffman_code_length
+            let input = meta.query_advice(input, Rotation::cur());
+            let code = decode::data_expr(&parts);
+            let word_begin = meta.query_advice(word_begin, Rotation::cur());
```

Comment:
```suggestion
            let q_word_begin = meta.query_fixed(word_begin, Rotation::cur());
```
When it's a fixed selector using `q_` prefix makes that more clear.